### PR TITLE
[PyOV]: Support direct comparison of PartialShape and Shape with list and tuple

### PR DESCRIPTION
**Pull request by hub-bla**
### Details:
 - Support direct comparison of PartialShape and Shape with list and tuple
 - Tests

I'm not entirely sure of the dynamic case and could use a hint. ;)
Thanks!

### Tickets:
 - 128813

### CODE REVIEW

File: src/bindings/python/tests/test_graph/test_core.py [hunk 1]
```diff
@@ -337,6 +337,30 @@ def test_partial_shape_equals():
     ps2 = PartialShape.dynamic(rank=3)
     assert ps1 == ps2
 
+    ps = PartialShape([1, 2, 3])
```
Comment (akuporos) [inline src/bindings/python/tests/test_graph/test_core.py:340]: Please, add also tests for `Shape`

File: src/bindings/python/tests/test_graph/test_core.py [hunk 2]
```diff
@@ -337,6 +345,30 @@ def test_partial_shape_equals():
...
+    list_ps = []
+    assert ps.get_min_shape() == tuple_ps
+    assert ps.get_min_shape() == list_ps
+
+    ps = PartialShape([Dimension(1), Dimension(2), Dimension(3), Dimension.dynamic()])
+    tuple_ps = (1, 2, 3, 0)
+    list_ps = [1, 2, 3, 0]
+    assert ps.get_min_shape() == tuple_ps
+    assert ps.get_min_shape() == list_ps
+
```
Comment (p-wysocki) [inline src/bindings/python/tests/test_graph/test_core.py:371]: Please add one last test: for `PartialShape` with `Dimension` of a range of values. 
Reference is the third `__init__` at: https://docs.openvino.ai/2022.3/api/ie_python_api/_autosummary/openvino.runtime.Dimension.html#openvino.runtime.Dimension.__init__

Interval `Dimension` means that it's dynamic, but the final value inside it has to be within the specified range.

**Review by p-wysocki — COMMENTED** [review]
Thank you for your contribution! One last thing to add, please let me know if you have any questions about it. When the tests are complete I'll approve. :)

**Review by p-wysocki — APPROVED** [review]

**Comment by hub-bla** [comment]
Did the linter get an update? I ran `python -m flake8 ./tests/ --config=setup.cfg` in `openvino/src/bindings/python` directory and it didn't throw me the error about variables called with only one char.

File: src/bindings/python/tests/test_graph/test_core.py [hunk 3]
```diff
@@ -319,6 +319,14 @@ def test_partial_shape_refinement():
     assert not ps2.relaxes(ps1)
 
 
+def test_shape_equals():
+    sh = Shape([1, 2, 3])
+    tuple_sh = (1, 2, 3)
+    list_sh = [1, 2, 3]
+    assert sh == tuple_sh
+    assert sh == list_sh
```
Comment (akuporos) [inline src/bindings/python/tests/test_graph/test_core.py:327]:
```suggestion
@pytest.mark.parametrize("shape", [[1, 2, 3], (1, 2, 3)])
def test_shape_equals(shape_to_compare):
    shape = Shape([1, 2, 3])
    assert shape == shape_to_compare
```

**Review by akuporos — APPROVED** [review]

**Comment by akuporos** [comment]
build_jenkins

File: src/bindings/python/src/pyopenvino/graph/partial_shape.cpp [hunk 4]
```diff
@@ -18,6 +18,17 @@
 
 namespace py = pybind11;
 
+template <typename T>
+bool compare_shape(const ov::PartialShape& a, const T& b) {
+    if (a.is_dynamic()) {
+        throw py::type_error("Cannot compare dynamic shape with " + std::string(typeid(b).name()));
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/partial_shape.cpp:24]: let's add one more test case that this message throws.
You may catch it like [here ](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/tests/test_graph/test_core.py#L99)

Comment (hub-bla) [reply src/bindings/python/src/pyopenvino/graph/partial_shape.cpp]: Done! I also changed `typeid(b).name()` for `py::str(py::type::of(b)))` so error message prints type in pythonic way like `<class 'tuple'>`. Thanks for pointing this out

**Comment by akuporos** [comment]
build_jenkins

**Comment by hub-bla** [comment]
> Did the linter get an update? I ran `python -m flake8 ./tests/ --config=setup.cfg` in `openvino/src/bindings/python` directory and it didn't throw me the error about variables called with only one char.

It happened again. Do I miss something?

**Comment by mlukasze** [comment]
build_jenkins

**Comment by akuporos** [comment]
build_jenkins

File: src/bindings/python/tests/test_graph/test_core.py [hunk 5]
```diff
@@ -337,6 +343,58 @@ def test_partial_shape_equals():
...
+            pass
+    assert (
+        "Cannot compare dynamic shape with <class 'tuple'>" in str(e.value)
+    )
+
+    with pytest.raises(TypeError) as e:
+        ps = PartialShape.dynamic()
+        list_ps = []
+        if ps == list_ps:
+            pass
```
Comment (akuporos) [inline src/bindings/python/tests/test_graph/test_core.py:393]:
```suggestion
        assert ps == list_ps
```

File: src/bindings/python/tests/test_graph/test_core.py [hunk 6]
```diff
@@ -337,6 +343,58 @@ def test_partial_shape_equals():
...
+    assert ps.get_min_shape() == tuple_ps_min
+    assert ps.get_max_shape() == tuple_ps_max
+    assert ps.get_min_shape() == list_ps_min
+    assert ps.get_max_shape() == list_ps_max
+
+    with pytest.raises(TypeError) as e:
+        ps = PartialShape.dynamic()
+        tuple_ps = ()
+        if ps == tuple_ps:
+            pass
```
Comment (akuporos) [inline src/bindings/python/tests/test_graph/test_core.py:384]:
```suggestion
        assert ps == tuple_ps
```

**Comment by akuporos** [comment]
> > Did the linter get an update? I ran `python -m flake8 ./tests/ --config=setup.cfg` in `openvino/src/bindings/python` directory and it didn't throw me the error
> 
> It happened again. Do I miss something?

Hello @hub-bla,

I'd guess that something wrong with flake8 version and its addons. I may just recommend your to reinstall our requiremets_test.txt

**Review by akuporos — APPROVED** [review]

**Comment by akuporos** [comment]
build_jenkins

**Pull request merged**